### PR TITLE
Gateway API: add support for TCPRoute

### LINKED ISSUE #5321: Gateway API: support TCPRoute
Blocked by #4960 / #5160 

Once the above work is merged, adding basic support for [TCPRoute](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TCPRoute) should be straight-forward.

### PR DESCRIPTION
**Pull request by skriss**
Closes #5321.

### CODE REVIEW

File: internal/dag/gatewayapi_processor.go [hunk 1]
```diff
@@ -600,6 +615,16 @@ func (p *GatewayAPIProcessor) getListenerRouteKinds(listener gatewayapi_v1beta1.
 			)
 			continue
 		}
+		if routeKind.Kind == KindTCPRoute && listener.Protocol != gatewayapi_v1beta1.TCPProtocolType {
```
Comment (skriss) [inline internal/dag/gatewayapi_processor.go:618]: I'm planning to add protocol=TLS + TCPRoute support in a separate PR.

**Review by sunjayBhatia — COMMENTED** [review]
nice! basically we add a new property on a listener separate from virtualhost since with plain tcp we can't do any sni/hostname routing and set up the tcp proxy filter accordingly 👍🏽 

wdyt about an e2e test for this? might be a while until there is a relevant one in upstream

File: internal/k8s/kind.go [hunk 2]
```diff
@@ -54,6 +54,8 @@ func KindOf(obj any) string {
 			return "GRPCRoute"
 		case *gatewayapi_v1alpha2.TLSRoute:
 			return "TLSRoute"
+		case *gatewayapi_v1alpha2.TCPRoute:
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/k8s/kind.go:57]: Added

File: internal/k8s/helpers.go [hunk 3]
```diff
@@ -113,10 +113,11 @@ func IsObjectEqual(old, new client.Object) (bool, error) {
 
 	case *gatewayapi_v1beta1.GatewayClass,
 		*gatewayapi_v1beta1.Gateway,
+		*gatewayapi_v1beta1.ReferenceGrant,
 		*gatewayapi_v1beta1.HTTPRoute,
 		*gatewayapi_v1alpha2.TLSRoute,
-		*gatewayapi_v1beta1.ReferenceGrant,
-		*gatewayapi_v1alpha2.GRPCRoute:
+		*gatewayapi_v1alpha2.GRPCRoute,
+		*gatewayapi_v1alpha2.TCPRoute:
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/k8s/helpers.go:120]: Added

File: test/e2e/gateway/tcproute_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		require.True(t, ok)
+		require.NotNil(t, route)
+
+		res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+			Condition: e2e.HasStatusCode(200),
+		})
+		assert.Truef(t, ok, "expected 200 response code, got %d", res.StatusCode)
+		assert.Equal(t, "echo", f.GetEchoResponseBody(res.Body).Service)
+
+		// Envoy is expected to add the "server: envoy" and
```
Comment (sunjayBhatia) [inline test/e2e/gateway/tcproute_test.go:68]: nice! i was gonna suggest something like this 👍🏽

**Review by sunjayBhatia — APPROVED** [review]

File: internal/dag/gatewayapi_processor.go (hunk 1 above)
Comment (skriss) [reply internal/dag/gatewayapi_processor.go:618]: ref. #5461

**Pull request merged**